File: app/src/main/java/com/jameskbride/eventbusmvpdemo/network/SecurityApiWrapper.java
package com.jameskbride.eventbusmvpdemo.network;

import com.jameskbride.eventbusmvpdemo.bus.SecurityErrorEvent;

import org.greenrobot.eventbus.EventBus;

import io.reactivex.functions.Consumer;
import retrofit2.HttpException;

public class SecurityApiWrapper<T extends Throwable> implements Consumer<T> {

    private EventBus eventBus;
    private Consumer consumer;

    public SecurityApiWrapper(EventBus eventBus, Consumer consumer) {
        this.eventBus = eventBus;
        this.consumer = consumer;
    }

    @Override
    public void accept(T t) throws Exception {
        if (t instanceof HttpException) {
            HttpException exception = (HttpException)t;
            if (401 == exception.code()) {
                eventBus.post(new SecurityErrorEvent());
            } else {
                consumer.accept(t);
            }
        } else {
            consumer.accept(t);
        }
    }
}

File: app/src/main/java/com/jameskbride/eventbusmvpdemo/network/SecurityApiWrapper.kt
package com.jameskbride.eventbusmvpdemo.network

import com.jameskbride.eventbusmvpdemo.bus.BusAware
import com.jameskbride.eventbusmvpdemo.bus.SecurityErrorEvent
import io.reactivex.functions.Consumer
import org.greenrobot.eventbus.EventBus
import retrofit2.HttpException

class SecurityApiWrapper<T : Throwable>(
        override val eventBus: EventBus,
        val consumer: Consumer<T>): Consumer<T>, BusAware {

    override fun accept(t: T) {
        if (t is HttpException) {
            val exception = t as HttpException
            if (401 == exception.code()) {
                eventBus.post(SecurityErrorEvent())
            } else {
                consumer.accept(t)
            }
        } else {
            consumer.accept(t)
        }
    }
}